In [1]:
import re

<font color='red'> **Trong phân mỗi phần đều có bài tập, các bạn cần điện code vào các dấu "... " trong mục "Bài tập"**</font>
## 1. THực hành với re.findall()

* `re.findall(pattern, string, flags=0)` là một hàm trong thư viện re của Python, được sử dụng để tìm tất cả các chuỗi con trong một chuỗi có khớp với một biểu thức chính quy.
    * `pattern`: biểu thức chính quy sẽ được sử dụng để xác định chuỗi tìm kiếm.
    * `string`: chuỗi cần tìm kiếm.
    * `flags`: là một đối số tùy chọn được sử dụng để thay đổi hành vi của biểu thức chính quy. VÍ dụ flags=re.IGNORECASE sẽ thể hiện không phân biệt chữ hoa và chữ thường khi thực hiện hàm
* Hàm findall() tìm kiếm toàn bộ chuỗi string để tìm tất cả các chuỗi con mà khớp với biểu thức chính quy pattern, và trả về một danh sách (list) chứa tất cả các chuỗi con tìm thấy. Nếu không tìm thấy, danh sách trả về sẽ rỗng.


In [ ]:
# Ví dụ 1:  tìm tất cả các chữ số
# \d: Thể hiện các chữ số từ 0 đến 9
text = "Học lập trình Python 123 #"
re.findall(r"[\d]", text)

['1', '2', '3']

In [2]:
# Ví dụ 2: Tìm tất cả các chữ cái in hoa
# A-Z: Thể hiện các chữ cái từ A -> Z trong bảng chữ cái
text = "Học lập trình Python 123 #"
re.findall(r"[A-Z]", text)

['H', 'P']

In [ ]:
# Ví dụ 3: tìm tất cả các từ cách nhau bởi khoảng trắng
# \b: Đánh dấu vị trí bắt đầu hoặc kết thúc của một từ.
# \w+: Tìm kiếm các ký tự chữ cái, số hoặc dấu gạch dưới liên tiếp.
text = "Học lập trình Python2 Python3 9 và 10 và 123 và 4567"
re.findall(r"\b\w+\b", text)

['Học', 'lập', 'trình', 'Python', '123']

### Bài tập
Cho đoạn text "Học lập trình Python 123 và 456", hãy thực hiện các yêu cầu:
* Tìm tất cả các chuỗi số liên tiếp từ 2 số trở lên: Kết quả: ['10', '123', '4567']
* Tìm tất cả các từ chứa chữ Python: Kết quả: ['Python2', 'Python3']


In [3]:
# Yêu cầu 1
text = "Học lập trình Python2 Python3 9 và 10 và 123 và 4567"
numbers = re.findall(r"\d{2,}",text)
print(numbers)

['10', '123', '4567']


In [4]:
#Yêu cầu 2
text = "Học lập trình Python2 Python3 9 và 10 và 123 và 4567"
words = re.findall(r"\bPython\w+",text)
print(words)

['Python2', 'Python3']


## 2. THực hành với re.search()

* `re.search(pattern, string, flags=0)`  là một hàm trong thư viện re của Python, được sử dụng để tìm kiếm một biểu thức chính quy trong một chuỗi.
    * `pattern`: biểu thức chính quy sẽ được sử dụng để xác định chuỗi cần tìm kiếm
    * `string`: chuỗi cần tìm kiếm.
    * `flags`: là một đối số tùy chọn được sử dụng để thay đổi hành vi của biểu thức chính quy. VÍ dụ flags=re.IGNORECASE sẽ thể hiện không phân biệt chữ hoa và chữ thường khi thực hiện hàm
* Hàm search() tìm kiếm biểu thức chính quy pattern trong chuỗi string và trả về một đối tượng match object nếu tìm thấy, hoặc None nếu không tìm thấy.

In [5]:
# Ví dụ 1: để tìm kiếm từ "Python" trong chuỗi "Học lập trình Python", chúng ta có thể sử dụng hàm search() như sau:
text = "Học lập trình Python Python"
match = re.search(r"Python", text)
if match:
    print("Tìm thấy từ 'Python' trong chuỗi:", match.group())
else:
    print("Không tìm thấy từ 'Python' trong chuỗi.")

Tìm thấy từ 'Python' trong chuỗi: Python


In [6]:
print(type(match))

<class 're.Match'>


### Bài tập
Cho chuỗi sau: "Học lập trình Python và R"
* Sử dụng hàm re.search() để tìm kiếm từ "Python" trong chuỗi trên. In ra vị trí bắt đầu và kết thúc của từ được tìm thấy.

* Sử dụng hàm re.search() để tìm kiếm từ "R" trong chuỗi trên. In ra vị trí bắt đầu và kết thúc của từ được tìm thấy.

* Sử dụng hàm re.search() để tìm kiếm từ "Java" trong chuỗi trên. Nếu không tìm thấy, in ra thông báo "Không tìm thấy từ Java trong chuỗi".

**Gợi ý:** Để lấy vị trí bắt đầu và kết thúc của đối tượng match, các bạn sẽ dùng phương thức match_object.start() và match_object.end()

In [7]:

text = "Học lập trình Python và R"

# Tìm kiếm từ "Python" trong chuỗi
match_python = re.search(r"Python",text)
if match_python:
    print(f"Tìm thấy từ Python từ vị trí {match_python.start()} đến {match_python.end()-1}")
else:
    print("Không tìm thấy từ Python trong chuỗi")


Tìm thấy từ Python từ vị trí 14 đến 19


In [8]:

# Tìm kiếm từ "R" trong chuỗi
match_R = re.search(r"R",text)
if match_R:
    print(f"Tìm thấy từ R từ vị trí {match_R.start()} đến {match_R.end()-1}")
else:
    print("Không tìm thấy từ R trong chuỗi")

Tìm thấy từ R từ vị trí 24 đến 24


In [9]:
# Tìm kiếm từ "Java" trong chuỗi
match_Java = re.search(r"Java",text)
if match_Java:
    print(f"Tìm thấy từ Java từ vị trí {match_Java.start()} đến {match_Java.end()-1}")
else:
    print("Không tìm thấy từ Java trong chuỗi")

Không tìm thấy từ Java trong chuỗi


## 3. THực hành với re.split()

* `re.split(pattern, string, maxsplit=0, flags=0)` là một hàm trong module re của Python được sử dụng để tách một chuỗi thành một danh sách các chuỗi con bằng cách sử dụng biểu thức chính quy để xác định điểm tách. Hàm này trả về một danh sách các chuỗi con.
    * `pattern`: Biểu thức chính quy sẽ được sử dụng để xác định điểm tách.
    * `string`: Chuỗi cần tách.
    * `maxsplit`: số lượng tối đa các chuỗi con được tạo ra. Nếu maxsplit không được cung cấp hoặc có giá trị bằng 0, tất cả các chuỗi con có thể được tạo ra.
    * `flags`: là một đối số tùy chọn được sử dụng để thay đổi hành vi của biểu thức chính quy. VÍ dụ flags=re.IGNORECASE sẽ thể hiện không phân biệt chữ hoa và chữ thường khi thực hiện hàm
    

In [10]:
# Ví dụ 1: Tách các từ dựa vào khoảng trắng
# \s+: Thể hiện 1 hoặc nhiều khoảng trắng
text = "Học lập trình   Python và         R"
words = re.split(r"\s+", text)
print(words)

['Học', 'lập', 'trình', 'Python', 'và', 'R']


### Bài tập
Cho chuỗi sau: "Học lập trình Python và R"
* Sử dụng hàm re.split() để tách chuỗi dưới đây thành danh sách các user:
    
    text = "john.doe@funix1.com, jane.doe@funix2.com, jsmith@funix3.com"

In [11]:
string = "john.doe@funix.com, jane.doe@funix.com, jsmith@funix.com"
email_list = re.split(r"[@]funix.com", string)
print(email_list)

['john.doe', ', jane.doe', ', jsmith', '']


## 4. Thực hành với re.sub()

* Hàm `re.sub(pattern, repl, string, count=0, flags=0)` trong Python được sử dụng để thay thế tất cả các mẫu phù hợp với biểu thức chính quy trong một chuỗi bằng một chuỗi khác. Cú pháp của hàm này như sau:
    * `pattern`: biểu thức chính quy cần tìm kiếm trong chuỗi.
    * `repl`: chuỗi thay thế cho mỗi mẫu phù hợp với pattern.
    * `string`: chuỗi gốc cần thực hiện thay thế.
    * `count (tùy chọn)`: số lần thực hiện thay thế. Nếu bằng 0 hoặc không được đặt, tất cả các mẫu phù hợp đều sẽ được thay thế.
    * `flags`: là một đối số tùy chọn được sử dụng để thay đổi hành vi của biểu thức chính quy. VÍ dụ flags=re.IGNORECASE sẽ thể hiện không phân biệt chữ hoa và chữ thường khi thực hiện hàm
    

In [12]:
# Ví dụ: Sử dụng hàm re.sub() để thay thế tất cả các khoảng trắng trong một chuỗi bằng dấu gạch ngang
string = "Hello   world!  How are  you?"
new_string = re.sub(r"\s+", "-", string)
print(new_string)

Hello-world!-How-are-you?


### Bài tập
* Hãy viết một hàm làm sạch dữ liệu bằng cách chỉ sử dụng hàm re.sub để thực hiện các công việc sau:
    * Thay thế tất cả các kí tự số bằng dấu sao (*).
    * Thay thế tất cả các từ bắt đầu bằng chữ cái thường bằng chữ cái in hoa.
    * Xóa tất cả các kí tự đặc biệt ($%^&*@#) trong chuỗi.

Sau khi viết xong, bạn hãy kiểm tra hàm vừa tạo bằng chuỗi string đã cho

In [13]:
def replace_chars(string):
    # Thay thế tất cả các kí tự số bằng dấu sao
    new_string = re.sub(r"\d", "*", string)

    # Thay thế tất cả các từ bắt đầu bằng chữ cái thường bằng chữ cái in hoa
    new_string = re.sub(r"\b[a-z]", lambda x: x.group().capitalize(), new_string)

    # Xóa tất cả các kí tự đặc biệt
    new_string = re.sub(r"[$%^&*@#]", "", new_string)

    return new_string

In [15]:
string = "The 1st example string! Please replace it with * instead. Thank you!"
new_string = replace_chars(string)
print(new_string)

The St Example String! Please Replace It With  Instead. Thank You!


## 5. Thực hành với re.match()
* `re.match(pattern, string, flags=0)` là một hàm trong module re của Python, dùng để kiểm tra xem một chuỗi có khớp với biểu thức chính quy được chỉ định hay không, và nếu có, trả về một đối tượng match, ngược lại trả về None.
    * `pattern` là biểu thức chính quy để tìm kiếm.
    * `string` là chuỗi cần kiểm tra.
    * `flags`: là một đối số tùy chọn được sử dụng để thay đổi hành vi của biểu thức chính quy. VÍ dụ flags=re.IGNORECASE sẽ thể hiện không phân biệt chữ hoa và chữ thường khi thực hiện hàm
    

In [16]:
#Ví dụ
# Kiểm tra chuỗi có bắt đầu bằng 'Hello' hay không
string1 = 'Hello, World!'
string2 = 'Hi, there!'
pattern = r'^Hello'

match1 = re.match(pattern, string1)
match2 = re.match(pattern, string2)

if match1:
    print("Match found in string1!")
else:
    print("No match found in string1.")

if match2:
    print("Match found in string2!")
else:
    print("No match found in string2.")

Match found in string1!
No match found in string2.


### Bài tập
* Viết một chương trình Python yêu cầu người dùng nhập vào một chuỗi và kiểm tra xem chuỗi đó có phải là số điện thoại hợp lệ hay không:
Một số điện thoại hợp lệ có định dạng như sau:

    * Bắt đầu bằng số 0.
    * Tiếp theo là hai hoặc ba chữ số (tùy thuộc vào quốc gia).
    * Sau đó là dấu gạch ngang (-) hoặc khoảng trắng.
    * Cuối cùng là một chuỗi gồm 7 hoặc 8 chữ số.
* Ví dụ các số điện thoại hợp lệ:
    * 0987654321
    * 012-3456789
    * 0123 456 789
    * 0908-123456
* Ví dụ các số điện thoại không hợp lệ:
    * 12345678
    * 01234
    * 09876-54321
    
Sau khi viết xong, bạn hãy kiểm tra hàm vừa tạo bằng phone_numbers đã cho

In [18]:
phone_numbers = ["0987654321", "987654321", "0987654321a", "0-123456789"]
# Biểu thức chính quy kiểm tra số điện thoại
pattern = r"^0\d{2,3}[-\s]?\d{7,8}\b"

for phone_number in phone_numbers:
    # Kiểm tra chuỗi có khớp với biểu thức chính quy hay không
    if re.match(pattern, phone_number):
        print("Valid phone number.")
    else:
        print("Invalid phone number.")

Valid phone number.
Invalid phone number.
Invalid phone number.
Invalid phone number.
